Add safe area and document coordinate systems

madsmtm opened the pull request:
Resolves https://github.com/rust-windowing/winit/issues/2308, by adding `Window::safe_area`, which describes the area of the surface that is unobstructed by notches, bezels etc. The drawing code in the examples have been updated to draw a star inside the safe area, and the plain background outside of it.

Also renamed `Window::inner_position` (introduced in https://github.com/rust-windowing/winit/pull/430) to `Window::surface_position`, and changed it to from screen coordinates to window coordinates, to better align how these coordinate systems work together.

Finally, I've added some SVG images and documentation to describe how these coordinate systems work together. The images should be auto-adjusting to the page theme when viewed on docs.rs. See also https://github.com/rust-windowing/winit/issues/3891, whether we use window coordinates or surface coordinates is currently a bit confusing.

Completing implementation on all platforms is tracked in https://github.com/rust-windowing/winit/issues/3910

Fixes https://github.com/rust-windowing/winit/issues/1122
Fixes https://github.com/rust-windowing/winit/issues/2308
Fixes https://github.com/rust-windowing/winit/issues/3742
Fixes https://github.com/rust-windowing/winit/issues/2066
Fixes https://github.com/rust-windowing/winit/issues/2347
Fixes https://github.com/rust-windowing/winit/discussions/2235

kchibisov commented on src/window.rs:
```diff
@@ -710,11 +706,29 @@ pub trait Window: AsAny + Send + Sync {
+    /// - **Wayland / Android / Orbital:** Unimplemented, returns `((0, 0), surface_size)`.
+    /// - **macOS:** This must be used when using `set_simple_fullscreen` to prevent overlapping the
+    ///   notch on newer devices.
+    fn safe_area(&self) -> (PhysicalPosition<u32>, PhysicalSize<u32>);
```
You can default implement it with `surface_position` and `surface_size`.

madsmtm replied:
Yeah, this would work:

```suggestion
    fn safe_area(&self) -> (PhysicalPosition<u32>, PhysicalSize<u32>) {
        ((0, 0).into(), self.surface_size())
    }
```

But not sure if it's a good idea, it might make backends forget to implement it? Idk., we haven't really figured out the story for when to default impl things on `Window`, and there's a lot of other places it could be done, so I'd rather avoid doing so in this PR, and instead discuss it separately.

kchibisov replied:
In this particular case, the API is not that wide spread, thus likely won't be implemented by many backends.

madsmtm replied:
I would like to keep things in this PR consistent with what we're already doing in the codebase. If we want to change how it's done, then it should go in another PR, because it applies to a lot of properties (`set_min_surface_size`, `surface_resize_increments`, `is_minimized` etc.), not just `safe_area`.

daxpedda replied:
I'm in favor of **not** default implementing anything.

kchibisov reviewed (commented):
Can macOS backend provide safe area when the transparent decorations are used? The ones where the buttons are over the main surface.

kchibisov commented on docs/res/coordinate-systems-desktop.svg:
positions are points, `outer_size` is `window_size`.

madsmtm replied:
`outer_size` is not `window_size` currently, and I'm not sure I want it to be? It's already defined on `Window`, so it seems a bit excessive to repeat "window". If anything, I'd rather call it just `size`, but that might lead people to use that over `surface_size`.

What do you mean by "positions are points"?

kchibisov replied:
>outer_size is not window_size currently, and I'm not sure I want it to be? It's already defined on Window, so it seems a bit excessive to repeat "window". If anything, I'd rather call it just size, but that might lead people to use that over surface_size.

it's a size with all the decorations, which is what we have and that's what we're reporting, so it's effectively a _window_ size.

If you think of window as a part without decorations, probably then not, but in the current state it's window_size from what I can say.

>What do you mean by "positions are points"?

I mean, you show them with lines as in they are areas, but you should just draw them as a point.

madsmtm replied:
> I mean, you show them with lines as in they are areas, but you should just draw them as a point.

Ah, yeah, but I wasn't sure how to represent what the point was relative to?

kchibisov replied:
maybe color coding, so different coordinate spaces will be differently colored, though, you already do so?

madsmtm replied:
I have pushed a new commit that updates the diagram to do something like this:
<img width="508" alt="screenshot" src="https://github.com/user-attachments/assets/206f0cec-74a5-4e97-850c-3bd43e4e1750">

kchibisov commented on src/window.rs:
```diff
@@ -710,11 +706,29 @@ pub trait Window: AsAny + Send + Sync {
+    /// ## Platform-specific
+    ///
+    /// - **Wayland / Android / Orbital:** Unimplemented, returns `((0, 0), surface_size)`.
+    /// - **macOS:** This must be used when using `set_simple_fullscreen` to prevent overlapping the
+    ///   notch on newer devices.
```
Shouldn't it generally be used when you move your window behind the notch, etc? Generally speaking, both fullscreen/simple fullscreen and window minimization could result in notch obscuring things.

Should we add a `SafeArea` event telling that it got updates or how it generally supposed to work on macOS? I'd assume that you should call it once you enter fullscreen/maximize/simple fullscreen states?

madsmtm replied:
Normal full screen modes place the content below the notch, so it's not an issue there.

> Should we add a `SafeArea` event

I think that'd be overcomplicating it, the user should _always_ handle `safe_area` in their rendering code, most of the time it'll just be a no-op. I included this comment for completeness, but perhaps it's more appropriate to only state it on `set_simple_fullscreen`.

kchibisov replied:
>I think that'd be overcomplicating it, the user should always handle safe_area in their rendering code, most of the time it'll just be a no-op.

The point was that you may change the state without changing the size, etc, though, it's a separate issue. Maybe it should just schedule redraw on macOS when you change the state of fullscreen, which is likely the case already.

madsmtm replied:
I think we should just schedule a redraw in the odd case where the safe area changes without the surface changing

madsmtm commented:
> Can macOS backend provide safe area when the transparent decorations are used? The ones where the buttons are over the main surface.

Didn't even know we had `WindowAttributesExtMacOS::with_titlebar_transparent`, but yeah, that should be possible.

I was kinda hoping I could define `safe_area` to not include CSDs, but seems like I'll have to update the docs to include it as well :/.

madsmtm commented:
I have ensured that `safe_area` includes the title bar when `with_titlebar_transparent` + `with_titlebar_hidden` is set on macOS

madsmtm commented:
I have implemented `surface_position` on all platforms now. Most of them were trivial, but **please test it on X11 and Windows**, as I am a bit unsure I got those correct. @notgull or others?

Additionally, I have implemented `safe_area` to simply return `((0, 0), surface_size)`. I will argue that a full implementation of this is not a requirement for getting this PR merged (otherwise it's never gonna get done), since rendering visible content in the entire surface, while wrong, is what the user did before anyhow.

kchibisov commented on src/window.rs:
```diff
@@ -710,11 +726,30 @@ pub trait Window: AsAny + Send + Sync {
+    ///
+    /// - **Android / Orbital / Wayland / Web / Windows / X11:** Unimplemented, returns `((0, 0),
+    ///   surface_size)`.
+    fn safe_area(&self) -> (PhysicalPosition<u32>, PhysicalSize<u32>);
```
The position and size is generally better, since it's easier to forward stuff like that into the e.g. `glViewport` stuff, since you generally want an _offset_ and not left`/right`.

daxpedda replied:
I think I prefer returning a type instead of a tuple.
WDYT?

madsmtm replied:
Have changed `safe_area` to return the insets instead

kchibisov commented on src/platform_impl/android/mod.rs:
```diff
@@ -796,8 +796,9 @@ impl CoreWindow for Window {
-    fn inner_position(&self) -> Result<PhysicalPosition<i32>, RequestError> {
-        Err(NotSupportedError::new("inner_position is not supported").into())
+    fn surface_position(&self) -> PhysicalPosition<i32> {
+        // FIXME: Complete this implementation
```
I'm not really a fan of having `FIXME` like that in the code, when the proper issue could be opened.

madsmtm replied:
I've opened https://github.com/rust-windowing/winit/issues/3910

kchibisov commented on src/window.rs:
```diff
@@ -590,41 +590,52 @@ pub trait Window: AsAny + Send + Sync {
-    /// - **Web:** Returns the top-left coordinates relative to the viewport. _Note: this returns
-    ///   the same value as [`Window::outer_position`]._
-    /// - **Android / Wayland:** Always returns [`RequestError::NotSupported`].
+    /// This _may_ be negative in the off case that the windowing system decided to make a surface
+    /// that is larger than your window.
```
I don't think it's how it's generally, the offset could be negative because it just decided to do so, it doesn't mean that it's bigger and the whole sentence could be dropped.

Or you can rephase it to say that it could be negative (and that's about it), but that's already clear from the `i32` type we're using.

madsmtm replied:
I just kinda wanted to make it explicit, but I'll try to reword.

kchibisov commented on src/lib.rs:
```diff
@@ -115,6 +115,41 @@
+//! between these three coordinate systems, although you should still strive to handle this, as
+//! they're still relevant in more niche area such as Mac Catalyst, or multi-tasking on tablets.
+//!
+//! There is, however, a different important concept: The "safe area". This can be accessed with
+//! [`Window::safe_area`], and describes a rectangle in the surface that is not obscured by notches,
+//! the status bar, and so on. You should be drawing your background and non-important content on
+//! the entire surface, but restrict important content (such as interactable UIs, text, etc.) to
+//! being drawn in the safe area.
```
This sounds a bit specific to mobile, while in reality it's not. Probably I just read it like that becouse `There's, however` part.

But generally, this has nothing to do with the coordinate systems, it should be in the `Drawing on the window` section.

madsmtm commented on examples/window.rs:
```diff
@@ -896,18 +900,55 @@ impl WindowState {
+        });
+
+        // Draw a star (without anti-aliasing) inside the safe area
+        let surface_size = self.window.surface_size();
+        let (origin, size) = self.window.safe_area();
+        let in_star = |x, y| -> bool {
+            // Shamelessly adapted from https://stackoverflow.com/a/2049593.
+            let sign = |p1: (i32, i32), p2: (i32, i32), p3: (i32, i32)| -> i32 {
+                (p1.0 - p3.0) * (p2.1 - p3.1) - (p2.0 - p3.0) * (p1.1 - p3.1)
+            };
+
+            let pt = (x as i32, y as i32);
+            let v1 = (0, size.height as i32 / 2);
+            let v2 = (size.width as i32 / 2, 0);
+            let v3 = (size.width as i32, size.height as i32 / 2);
+            let v4 = (size.width as i32 / 2, size.height as i32);
+
+            let d1 = sign(pt, v1, v2);
+            let d2 = sign(pt, v2, v3);
+            let d3 = sign(pt, v3, v4);
+            let d4 = sign(pt, v4, v1);
+
+            let has_neg = (d1 < 0) || (d2 < 0) || (d3 < 0) || (d4 < 0);
+            let has_pos = (d1 > 0) || (d2 > 0) || (d3 > 0) || (d4 > 0);
 
-        let color = match self.theme {
-            Theme::Light => WHITE,
-            Theme::Dark => DARK_GRAY,
+            !(has_neg && has_pos)
         };
+        for y in 0..size.height {
+            for x in 0..size.width {
+                if in_star(x, y) {
+                    let index = (origin.y + y) * surface_size.width + (origin.x + x);
+                    buffer[index as usize] = match self.theme {
+                        Theme::Light => 0xffe8e8e8, // Light gray
+                        Theme::Dark => 0xff525252,  // Medium gray
+                    };
+                }
+            }
```
The nice thing about the star is that then you can very clearly see whether it lines up with the edges or not.

For example on Mac Catalyst, the surface is actually obscured by the title bar, but I couldn't see this just by colouring the unsafe part differently.

madsmtm commented on src/platform_impl/linux/wayland/window/mod.rs:
```diff
@@ -338,6 +337,11 @@ impl CoreWindow for Window {
     }
 
+    fn safe_area(&self) -> (PhysicalPosition<u32>, PhysicalSize<u32>) {
+        // FIXME: Include CSDs drawn by Winit
```
Hmm, should we then update `surface_position` to include the CSDs drawn by Winit?

kchibisov replied:
On Wayland, I think it won't change anything since CSDs are in negative part of the window, so the surface is still at (0,0), it's just CSDs are in (-border, -header), so don't have to do anything, since it's in the (0,0).

madsmtm replied:
Wow, that's wild! How does the compositor allow you to create and draw surfaces outside the bounds that you are otherwise restricted to?

kchibisov commented on src/lib.rs:
```diff
@@ -138,17 +146,13 @@
-//! the status bar, and so on. You should be drawing your background and non-important content on
-//! the entire surface, but restrict important content (such as interactable UIs, text, etc.) to
-//! being drawn in the safe area.
+//! Note that the safe area (discussed above) is especially important here, since windows on mobile
+//! are often full screen, and often the device has notches (as illustrated in the image below).
```
Still not sure how it applies to coordinate systems, since it doesn't change anything about them, because safe area doesn't alter how they work in any way and just says that you have _void areas_ in your _rendering_, but coordinate systems are more about directions.

madsmtm replied:
How about:
```suggestion
//! This is illustrated in the image below, along with the safe area since it's often relevant on
//! mobile.
```

?

kchibisov replied:
I guess it's fine.

kchibisov commented:
Just thinking more about it and that you should generally always call `safe_area` inside your rendering is that, I actually can not do that to achieve good behavior, since the only place where I can call it to achieve what I want (sync resize of terminal content and window) is during the `SurfaceResized` event, since I need to alter the terminal viewport size which I'll send via signal downstream during it, otherwise if I do that unconditionally in rendering part, I'd need more frames.

So `RedrawRequsted` is not the greatest place for that, since you generally know that you got resized way before that, and that. It also doesn't mean that you have to redraw, so I'd probably try to put it into its own event, then the users can decide whether they want to redraw or not, instead of us doing so for them, what do you think?

madsmtm commented:
Hmm, if you set the terminal viewport to only be the safe area, what does Alacritty intend to draw in the unsafe area, then?

kchibisov commented:
>Hmm, if you set the terminal viewport to only be the safe area, what does Alacritty intend to draw in the unsafe area, then?

just padding with background color, it's just the safe area changes the size of the content, which we should forward down the line for us? It's not really an issue for regular GUIs, I guess.

madsmtm commented:
Okay, so if it's something that _only_ affects Alacritty, then I'd argue that we don't need the event, since Alacritty only works on desktop anyhow, and I'm fine for now with guaranteeing for that case that on macOS, a change to the safe area will always correspond with a `SurfaceResized` ; see https://github.com/rust-windowing/winit/pull/3890/commits/62a95a5a946420082655d64dcdc2b888dbb13bfe.

I'm not in general against an event for it, but it's kinda cumbersome to set up, and I'd like to see a better use-case for it - and besides, it can be added after this PR.

kchibisov commented:
I don't think that the issue is _specific_ to alacritty though, it's specific to any application that computes layout and then decides whether to render or not.

Re-running your layout calculation each time during rendering is not that great, unless you're _egui_ or something, but for retained UIs they probably want to react to state a bit _earlier_.

I'd also not mention alacritty in the source, since it makes it sound like _its_ problem, where in reality any program that forwards sizes to some external _process_ will work the same, like vnc, etc.

In general though, I'd assume that on mobile it's not really an issue since you have rotation event, when it'll change the safe area and you can just call it once, since it doesn't change over time.

The thing about safe area though, is that it can not really change for the given monitor, because it's a _hardware_ property, what changes is the state of winit decorations and the state where this safe area should be honored, so if talking about event, you don't have to hook into the actual change to the bits you're quering for safe area, but to send this event from places where system wants you to honor/stop honoring the safe area.

In case of macOS, it'll be when you enter simple fullscreen mode, and in case of ios/android you can just always send it once you open your application or rotation happens (if rotation changes coordinate system).

So I don't see it being cumbersome to setup, since the area itself can not change.

madsmtm commented:
I'm mostly against such an event because if we want it to be useful, we also have to emit it every time we do `SurfaceResized`, on every platform, and _that's_ cumbersome.

Again, I believe it does not have to be handled in this PR, so I've opened https://github.com/rust-windowing/winit/issues/3911, let's discuss it there instead.

kchibisov commented:
>I'm mostly against such an event because if we want it to be useful, we also have to emit it every time we do SurfaceResized, on every platform, and that's cumbersome.

It doesn't have to be synced though, you should emit it only when the area changes, resize doesn't mean that it changed, because it may not apply at all.

I won't block on event, since it doesn't really matter for the majority of the clients though.

daxpedda commented on src/lib.rs:
```diff
@@ -115,6 +115,45 @@
 //! [`visible` set to `false`][crate::window::WindowAttributes::with_visible] and explicitly make
 //! the window visible only once you're ready to render into it.
 //!
+//! There is another important concept you need to know about when drawing: the "safe area". This
+//! can be accessed with [`Window::safe_area`], and describes a rectangle in the surface that is not
+//! obscured by notches, the status bar, and so on. You should be drawing your background and
+//! non-important content on the entire surface, but restrict important content (such as
+//! interactable UIs, text, etc.) to only being drawn inside the safe area.
+//!
+//! [`Window::safe_area`]: crate::window::Window::safe_area
+//!
+//! # Coordinate systems
+//!
+//! Windowing systems use many different coordinate systems, and this is reflected in Winit as well;
+//! there are "desktop coordinates", which is the coordinates of a window or monitor relative to the
+//! desktop at large, "window coordinates" which is the coordinates of the surface, relative to the
+//! window, and finally "surface coordinates", which is the coordinates relative to the drawn
+//! surface. All of these coordinates are relative to the top-left corner of their respective
+//! origin.
+//!
+//! Most of the functionality in Winit works with surface coordinates, so usually you only need to
+//! concern yourself with those. In case you need to convert to some other coordinate system, Winit
+//! provides [`Window::surface_position`] and [`Window::surface_size`] to describe the surface's
+//! location in window coordinates, and Winit provides [`Window::outer_position`] and
+//! [`Window::outer_size`] to describe the window's location in desktop coordinates. Using these
+//! methods, you should be able to convert a position in one coordinate system to another.
+//!
+//! An overview of how these four methods fit together can be seen in the image below:
+#![doc = concat!("\n\n", include_str!("../docs/res/coordinate-systems-desktop.svg"), "\n\n")] // Rustfmt removes \n, re-add them
+//! On mobile, the situation is usually a bit different; because of the smaller screen space,
+//! windows usually fill the whole screen at a time, and as such there is _rarely_ a difference
+//! between these three coordinate systems, although you should still strive to handle this, as
+//! they're still relevant in more niche area such as Mac Catalyst, or multi-tasking on tablets.
+//!
+//! This is illustrated in the image below, along with the safe area since it's often relevant on
+//! mobile.
+#![doc = concat!("\n\n", include_str!("../docs/res/coordinate-systems-mobile.svg"), "\n\n")] // Rustfmt removes \n, re-add them
```
```suggestion
#![doc = concat!("\n\n", include_str!("../docs/res/coordinate-systems-mobile.svg"), "\n\n")] // Rustfmt removes \n, adding them like this works around that.
```
"re-add them" sounded like a suggestion to me to do so every time we format.

daxpedda replied:
Amazing ... I hope the rest of our documentation catches up to this level of quality.
Thank you so much @madsmtm!

daxpedda commented on src/window.rs:
```diff
@@ -654,16 +664,21 @@ pub trait Window: AsAny + Send + Sync {
+    ///
+    /// Note that to ensure that your content is not obscured by things such as notches or the title
+    /// bar, you will likely want to only draw important content inside a specific area of the
+    /// surface, see [`safe_area`] for details.
```
```suggestion
    /// surface, see [`safe_area()`] for details.
```

daxpedda reviewed (approved):
This looks great!
I will work on the Web implementation soon.

daxpedda reviewed (commented):
Does the safe area account for the virtual keyboard as well on iOS?

nixpulvis commented:
https://developer.apple.com/design/human-interface-guidelines/layout#iOS-keyboard-layout-guide

daxpedda commented on src/window.rs:
```diff
@@ -710,11 +725,30 @@ pub trait Window: AsAny + Send + Sync {
+    /// of the surface, and the size extending downwards to the right. The area will not extend
+    /// beyond [the bounds of the surface][Window::surface_size].
+    ///
+    /// If the entire content of the surface is visible, this returns `((0, 0), surface_size)`.
```
I think this should note that it doesn't account for occluded areas.

I briefly thought about if we should account for occluded areas, but users need both separately.
Because safe area can define how you layout thing in your application, but occluded should not.

kchibisov replied:
Safe area is area without occlusion from hardware elements. Like notches, etc. It's nothing more that that. It's also the case for e.g. macOS decorations in some cases, but it's _winit specific_.

Regular occlusion by other applications is not it.

Virtual keyboards just resize your entire application on some platforms, so it should be dealt with on case by case basis. On Wayland, you're resized to not draw inside the virtual keyboard exclusive area. On ios/android, probably something else.

madsmtm replied:
Added comment to this effect

daxpedda commented:
I implemented Web as well now.

However there is a bug when not used with `viewport-fit=cover` that I can't figure out on my own.
Basically Safari adds margins/padding to both sides of the viewport when holding an iPhone with a notch sideways (even though the notch is only on one side). I'm unable to detect/read this padding/margin, so the calculations are all off.

This could also be happening on Android, unfortunately I can't test it because I don't have an Android device with a notch.

daxpedda commented:
So I spent some more time thinking about the Web implementation.
I have come to the conclusion that implementing this for Web is kind of impossible with the current description of the API.

AFAIK on iOS you can't move the window partly off-screen, but on Web it is possible to do so.
The `Window` on Web can be anywhere in the viewport, so what do we report when the user scrolls?
If we start accounting for all that, its looks more like we are representing occlusion and not the safe area.
We could start defining the safe area as the safe area without any scrolling and movement (?) applied.
What do we do if the `Window` is bigger then the viewport? If we calculate the safe area by removing the part that's bigger then the viewport we are back at occlusion again.

A practical example:
If the canvas sits in a webpage that has text around it and can be scrolled, the safe area might be used to calculate layout. If the user starts scrolling down, the layout would suddenly starts to squish together because a part of the surface is outside the safe area.
See [this Bevy example](https://bevyengine.org/examples/ui-user-interface/borders) where this exact same scenario would cause problems.

---

It seems to me that the safe area is only useful if the `Window` is as big (or smaller?) than the screen and, in the case of Web, the surface can't move around in the Web `Window` through scrolling.
Therefor I suggest that we define the safe area on the `MonitorHandle` instead of on the `Window`.
To calculate the safe surface area, we can keep the current function, but return an `Option`, which would be `None` if our definition of "safe area" is not applicable. E.g. the `Window` is not as big as the screen.

I'm unsure how to implement this on Web. Probably if the canvas is not as big as the viewport or the `Document` is bigger than the viewport (ergo if it causes any scrolling bars), we should return `None`.

We might want to define a new concept like "visible surface area", which exposes the truly visible area to the user, which would account for everything: safe area, occlusion, virtual keyboard and so on.
This can be implemented on Web with the [Intersection Observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API).

kchibisov commented:
Safe area matters for fullscreen window to not  draw behind the notch, for the rest it's not used. It's also _static_ to display coordinates (most of the time).

Like the whole purpose here is for e.g. alacritty offset it viewport to not put anything behind notch on macOS during fullscreen, nothing more, and yes, it's called safe area.

madsmtm commented:
@daxpedda and I talked about this today, and a rough conclusion was to calculate the safe area independently of the scroll position of the page, i.e. in a sort of "absolute" position of the entire document.

MarijnS95 commented:
I haven't exactly followed this PR, but how is this "surface inside window" representation going to come together with #3942 and all future "subsurface" changes that we're going to make? Every individual surface having its position relative to the parent surface and eventually relative to the window "area" that we define here?

madsmtm commented:
> I haven't exactly followed this PR, but how is this "surface inside window" representation going to come together with #3942 and all future "subsurface" changes that we're going to make? Every individual surface having its position relative to the parent surface and eventually relative to the window "area" that we define here?

Not sure yet, but I think it makes sense to allow querying the safe area relative to each surface.

madsmtm commented:
I have changed `fn safe_area` to return the insets (`dpi::PhysicalInsets { top, left, bottom, right }`) instead, since I belive that will be more forwards-compatible with https://github.com/rust-windowing/winit/issues/3911 and other improvements (discussed on Matrix today to add insets for each type of notch).

Remaining is for @daxpedda to fix the web wrt. scrolling, as I wrote shortly about in https://github.com/rust-windowing/winit/pull/3890#issuecomment-2438243038.

kchibisov commented on docs/res/coordinate-systems-desktop.svg:
This image looks cut on the left size when I look into it, is it the same for you?

madsmtm replied:
The "window" in the image is intentionally off-center, because there is more text on the right than on the left.

kchibisov replied:
![image](https://github.com/user-attachments/assets/effadae4-ebde-4918-b517-479f08c543d4)

At least firefox doesn't let me look into that `s` at all.

kchibisov replied:
it's the same everywhere for me, maybe firefox issue.

madsmtm replied:
Ahh, I think it's because I'm using the font `monospace`, which happens to be larger on your device.

kchibisov replied:
The font on the picture is not monospace though, it's proportional (look at `t` and `i`).

madsmtm replied:
How's it now?

kchibisov replied:
Now it's ok.

kchibisov commented on examples/window.rs:
```diff
@@ -917,18 +921,63 @@ impl WindowState {
+        let in_star = |x, y| -> bool {
+            // Shamelessly adapted from https://stackoverflow.com/a/2049593.
+            let sign = |p1: (i32, i32), p2: (i32, i32), p3: (i32, i32)| -> i32 {
+                (p1.0 - p3.0) * (p2.1 - p3.1) - (p2.0 - p3.0) * (p1.1 - p3.1)
+            };
+
+            let pt = (x as i32, y as i32);
+            let v1 = (0, size.height as i32 / 2);
+            let v2 = (size.width as i32 / 2, 0);
+            let v3 = (size.width as i32, size.height as i32 / 2);
+            let v4 = (size.width as i32 / 2, size.height as i32);
+
+            let d1 = sign(pt, v1, v2);
+            let d2 = sign(pt, v2, v3);
+            let d3 = sign(pt, v3, v4);
+            let d4 = sign(pt, v4, v1);
 
-        let color = match self.theme {
-            Theme::Light => WHITE,
-            Theme::Dark => DARK_GRAY,
+            let has_neg = (d1 < 0) || (d2 < 0) || (d3 < 0) || (d4 < 0);
+            let has_pos = (d1 > 0) || (d2 > 0) || (d3 > 0) || (d4 > 0);
+
+            !(has_neg && has_pos)
         };
```
Can't we just draw safe are in a different color, it should be much simpler than that, no?

kchibisov replied:
I just don't want to add even more code to this example.

madsmtm commented:
I've merged this now, though note that it's only really implemented for macOS and iOS (and Wayland, because there's no protocol for it). Web is implemented, though as noted above, it needs to be changed.

All of this remaining work (fixing the web implementation, implementing properly on the other platforms) is tracked in https://github.com/rust-windowing/winit/issues/3910.
